luarules/gadgets/cmd_nano_range_check.lua

```diff
@@ -0,0 +1,148 @@
+function gadget:GetInfo()
+	return {
+		name = "Construction Turrets Range Check",
+		desc = "Stops construction turrets from getting assigned to guards, repair, reclaim and attacks out of reach.",
+		author = "Nehroz",
+		date = "2024.12.01",
+		layer = 0,
+		enabled = true,
+	}
+end
+
+if not gadgetHandler:IsSyncedCode() then return end
+Spring.SetLogSectionFilterLevel(gadget:GetInfo().name, LOG.INFO) -- WARN Remove in final version.
+Spring.Log(gadget:GetInfo().name, LOG.INFO,"Unsync Load.")
+
+
+--- key is the nano's ID, followed by a Set-like with the unitID of target currently inside the nano's range.
+--- Used for CommandArray manipulation
+---@type {[number] : {number : boolean}}
+local trackingTable = {}
+
+local constructionTurretsDefs = {}
+for unitDefID, unitDef in ipairs(UnitDefs) do
+	if unitDef.isFactory == false and unitDef.isStaticBuilder then
+		constructionTurretsDefs[unitDefID] = {
+			maxBuildDistance = unitDef.buildDistance + unitDef.radius,
+			buildRange3D = unitDef.buildRange3D,
+		}
+	end
+end
+
+local function isEmptyTable(tbl)
+	for _ in pairs(tbl) do
+		return false
+	end
+	return true
+end
+
+-- only used for Logging; remove with it.
+local function getTableSize(tbl)
+	local i = 0
+	for _ in pairs(tbl) do
+		i = i + 1
+	end
+	return i
+end
+
+function gadget:Initialize()
+	gadgetHandler:RegisterAllowCommand(CMD.REPAIR)
+	gadgetHandler:RegisterAllowCommand(CMD.GUARD)
+	gadgetHandler:RegisterAllowCommand(CMD.RECLAIM)
+end
+
+local gate = false
+function gadget:GameFrame(frame)
+	local removes = 0
+	local inside = 0
+	local nanosRun = 0
+	local unitsCalced = 0
+
+	if frame % 60 > 10 then return end -- run 10 frames, skip 50; allow LuaUser.cpp to catch up. running full every 1s
+	local pointer = frame % 10
+	for nanoID in pairs(trackingTable) do
+		if pointer % 10 == 0 then
+			local nanoDefID = Spring.GetUnitDefID(nanoID)
+			local maxDistance = constructionTurretsDefs[nanoDefID].maxBuildDistance
+			local commands = Spring.GetUnitCommands(nanoID, -1)
+
+			for i=#commands, 1, -1 do
+				local unitID = commands[i].params[1]
+				if Spring.ValidUnitID(unitID) then
+					local distance = Spring.GetUnitSeparation(nanoID, unitID, constructionTurretsDefs[nanoDefID].buildRange3D, false)
+					if distance < maxDistance then -- Inside range
+						trackingTable[nanoID][unitID] = true
```

Comment:
I think the second level inside the tracking table is useless since the command queue is already tracking them here. imo just remove when distance is greater than maxDistance, and make trackingTable just one level (`trackingTable[nanoID] = true/false`).

Reply:
This is done for gameplay reasons, so you can pre-issue a command targeting a mobile unit before it gets into range, but the order is removed once the unit leaves range.

Reply:
put that in a --comment